### PR TITLE
initial commit for changes to support database as backend for audit logs

### PR DESCRIPTION
**Pull request by niravparikh05**
### What does this PR change?

- introduces database as an option to store audit logs

### Does the PR depend on any other PRs or Issues? If yes, please list them.

- https://github.com/paralus/helm-charts/pull/57

### Checklist

I confirm, that I have...

- [x] Read and followed the contributing guide in `CONTRIBUTING.md`
- [ ] Added tests for this PR
- [x] Formatted the code using `go fmt` (if applicable)
- [ ] Updated [documentation on the Paralus docs site](https://github.com/paralus/website/blob/main/docs) (if applicable)
- [ ] Updated `CHANGELOG.md`

### CODE REVIEW

File: pkg/service/relay_audit_db.go [hunk 1]
```diff
@@ -0,0 +1,99 @@
...
+			},
+			GroupByMethod: &auditv1.AggregatorGroup{
+				Buckets: buildAggregators(methodAggr),
+			},
+		},
+		Hits: &auditv1.Hits{Hits: buildDataSource(auditLogs)},
+	}
+
+	var resMap map[string]interface{}
+	data, _ := json.Marshal(response)
```
Comment (akshay196) [inline pkg/service/relay_audit_db.go:91]: We should handle the error in Marshaling.

**Review by akshay196 — CHANGES_REQUESTED** [review]
Few minor changes.

File: main.go [hunk 2]
```diff
@@ -359,34 +364,69 @@ func setup() {
...
+		}
+		rcs, err = service.NewAuditLogElasticSearchService(elasticSearchUrl, relayCommandsESIndexPrefix+"-*", "RelayCommand API: ")
+		if err != nil {
+			if dev && strings.Contains(err.Error(), "connect: connection refused") {
+				_log.Warn("unable to create auditLog service:", err)
+			} else {
+				_log.Fatalw("unable to create auditLog service", "error", err)
+			}
+		}
+	default:
```
[the remark this answers is not in the document]
Comment (niravparikh05) [reply main.go:428]: added a log warn, we continue without auditing ability as we currently do.

File: pkg/service/relay_audit_db.go (hunk 1 above)
Comment (niravparikh05) [reply pkg/service/relay_audit_db.go]: done

**Comment by akshay196** [comment]
@niravparikh05 I have setup this along with helm-chart PR and set `AUDIT_LOG_STORAGE=database`.
Dashboard is receiving an empty auditlog response even though the audit_logs table rows.

Table `audit_logs`:
```
select * from audit_logs;
;
     tag     |            time            |                                                                                                        
                                                                                                                                                   
                      data                                                                                                                         
                                                                                                                                                   
      
-------------+----------------------------+--------------------------------------------------------------------------------------------------------
---------------------------------------------------------------------------------------------------------------------------------------------------
---------------------------------------------------------------------------------------------------------------------------------------------------
---------------------------------------------------------------------------------------------------------------------------------------------------
------
 system      | 2022-11-23 06:15:42.44675  | {"date": 1669184142.44675, "type": "cluster.create.success", "actor": {"type": "USER", "groups": ["Orga
nization Admins", "All Local Users"], "account": {"username": "admin@paralus.local"}}, "client": {"ip": "10.244.0.9:39474", "host": "console.paralu
s.local", "type": "BROWSER", "user_agent": "Mozilla/5.0 (X11; Linux x86_64; rv:106.0) Gecko/20100101 Firefox/106.0"}, "detail": {"meta": {"cluster_
name": "import"}, "message": "Cluster import created"}, "origin": "core", "portal": "OPS", "project": "default", "version": "1.0", "category": "AUD
IT"}
 system      | 2022-11-23 06:15:45.365647 | {"date": 1669184145.365647, "type": "cluster.update.success", "actor": {"type": "USER", "groups": ["Org
anization Admins", "All Local Users"], "account": {"username": "admin@paralus.local"}}, "client": {"ip": "10.244.0.9:39474", "host": "console.paral
us.local", "type": "BROWSER", "user_agent": "Mozilla/5.0 (X11; Linux x86_64; rv:106.0) Gecko/20100101 Firefox/106.0"}, "detail": {"meta": {"cluster
_name": "import"}, "message": "Cluster import updated"}, "origin": "core", "portal": "OPS", "project": "default", "version": "1.0", "category": "AU
DIT"}
 kubectl_cmd | 2022-11-23 06:17:30.085495 | {"date": 1669184250.085495, "type": "kubectl.command.detail", "actor": {"type": "USER", "groups": ["Org
anization Admins", "All Local Users"], "account": {"username": "admin@paralus.local"}}, "client": {"ip": "172.18.0.1", "host": "console.paralus.loc
al", "type": "BROWSER", "user_agent": "Mozilla/5.0 (X11; Linux x86_64; rv:106.0) Gecko/20100101 Firefox/106.0"}, "detail": {"meta": {"cluster_name"
: "import"}, "message": "kubectl get pod"}, "origin": "cluster", "portal": "ADMIN", "project": "default", "version": "1.0", "category": "AUDIT"}
 kubectl_api | 2022-11-23 06:17:26.538    | {"d": 0.582536346, "k": "", "m": "GET", "n": "", "o": "cce8e594-a436-4d7e-ac5d-c5097e7723ca", "p": "ec9
--More--
```
![Screenshot from 2022-11-23 12-00-06](https://user-images.githubusercontent.com/16691314/203484866-8e4261e9-7db6-4f26-89f4-cc71cd277c50.png)
![Screenshot from 2022-11-23 11-59-55](https://user-images.githubusercontent.com/16691314/203484870-a6f2ca4c-676f-441f-a7a8-b0f871605acc.png)

Do we need any changes in Dashboard?

**Comment by niravparikh05** [comment]
hey @akshay196 thanks for pointing it out. Current code was referring to timestamp, ts json attributes from data column in auditlogs which is now reconfigured with time column hence query was not returning anything.

Have modified the query and updated the pr.

image

**Comment by akshay196** [comment]
@niravparikh05 Tested locally, auditing to databaase is functionally working.

**Review by akshay196 — APPROVED** [review]
Tested locally with Nirav. Working fine.

**Pull request merged**